rust-crates/symblib/src/disas.rs

```diff
@@ -0,0 +1,242 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+//! Minimal instruction disassembler implementation.
+
+use crate::{AnyError, VirtAddr};
+use fallible_iterator::FallibleIterator;
+use std::iter;
+
+/// Errors that can occur during instruction decoding.
+#[non_exhaustive]
+#[allow(missing_docs)]
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("Not enough bytes to decode the full instruction at {:#08X}", .0)]
+    TruncatedInstruction(VirtAddr),
+
+    #[error("Bytes at {:#08X} do not form a valid instruction", .0)]
+    InvalidInstruction(VirtAddr),
+
+    #[error(transparent)]
+    Other(AnyError),
+}
+
+/// Trait for instruction decoders.
+pub trait InstrDecoder {
+    /// Decode one instruction and return information.
+    fn decode(&self, addr: VirtAddr, buf: &[u8]) -> Result<InstrInfo, Error>;
+}
+
+/// Information about an instruction.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct InstrInfo {
+    /// Virtual addresss of the instruction.
+    pub addr: VirtAddr,
+
+    /// Whether the instruction is a call or syscall.
+    pub is_call: bool,
+
+    /// Length of the instruction, in bytes.
+    pub length: u8,
+}
+
+/// AMD64 (X86-64) instruction decoder.
+///
+/// Currently implemented using the Zydis library.
+#[derive(Debug)]
+pub struct Amd64InstrDecoder(zydis::Decoder);
```

Comment:
Have you considered implementing small subset of the disassembler in a rust? 
Asking because of security reasons. I would be cautious to run a disassembler written in C on potentially untrusted user data, especially if running as root during ontarget symbolization.

It would be nice at least to document somehow, that the amd64 retpad extractor is using C library for security visibility.

Reply:
Other than with ARM64, it's unfortunately not possible to implement a subset of the disassembler for AMD64. It's a variable length instruction set, so you need to be able to correctly decode each instruction in the program in order to find the start of the next one.